File: impls/tsbun/step2_eval.ts
import { read_str, read_str_antlr } from "./reader.ts";
import { pr_str, pr_str_antlr } from "./printer.ts";
import {Env} from "./env.ts";

const rootEnv = new Env();
rootEnv.set("+", (a : number, b : number) => a + b);
rootEnv.set("-", (a : number, b : number) => a - b);
rootEnv.set("*", (a : number, b : number) => a * b);
rootEnv.set("/", (a : number, b : number) => a / b);

const eval_ast = (ast: any, env: any): any => {
  return ast.eval(env);
}
const READ = (str: string): any => read_str_antlr(str);
const EVAL = (ast: any, _env?: any): any => eval_ast(ast, _env);
const PRINT = (exp: any): string => pr_str_antlr(exp)
const rep = (str: string): string => PRINT(EVAL(READ(str), rootEnv));

const repl = async () => {
  const prompt = "user> ";
  process.stdout.write(prompt);
  for await (const line of console) {
    if (line.length === 0) {
      break;
    }
    try {
      process.stdout.write(rep(line));
      process.stdout.write("\n");
    } catch (e) {
      const err: Error = e;
      process.stdout.write(err.message);
      process.stdout.write("\n");
    }
    process.stdout.write(prompt);
  }
}
repl();

File: impls/elixir2/lib/mix/tasks/step2_eval.ex

defmodule Mix.Tasks.Step2Eval do
  def run(_), do: loop()

  defp loop do
    IO.write(:stdio, "user> ")
    IO.read(:stdio, :line)
      |> read_eval_print
      |> IO.puts

    loop()
  end

  defp read(input) do
    Mal.Reader.read_str(input)
  end

  defp eval_ast({:symbol, ast}, env) do
    case env[ast] do
      nil -> {:error, "Symbol not found: #{ast}"}
      val -> val
    end
  end



  defp eval_ast({:list, [fnSym | rest]}, env) do
    # This is a special case for now
    thefn = eval_ast(fnSym, env)
    args = rest |> Enum.map(fn(x) -> eval_ast(x, env) end)
    apply(thefn, args)
  end
  defp eval_ast([] = l, _), do: l
  defp eval_ast(base, _) when is_number(base), do: base

  defp eval(input, env) do
    eval_ast(input, env)
  end

  defp print(input) do
    Mal.Printer.print_str(input)
  end

  defp read_eval_print(:eof), do: exit(:normal)
  defp read_eval_print(line) do
    env = %{
      "+" => fn (a, b) -> a + b end,
      "-" => fn (a, b) -> a - b end,
      "*" => fn (a, b) -> a * b end,
      "/" => fn (a, b) -> a / b end,
    }
    read(line)
      |> (&eval(&1, env)).()
      |> print
  catch
    {:error, msg} -> "Error: #{msg}"
  end

end
